File: kernel/arch/x86/memory/pm.c
// x86 physical memory allocator.

#include <menix/arch.h>
#include <menix/common.h>
#include <menix/log.h>
#include <menix/memory/pm.h>
#include <menix/thread/spin.h>
#include <menix/util/bitmap.h>

#include <string.h>

static SpinLock lock;
static BitMap bit_map = NULL;		// This bitmap stores whether a page is in use or not.
static void* phys_addr = NULL;		// Memory mapped lower 4GiB physical memory. This is only used to store the bitmap.
static usize num_pages = 0;			// Total amount of available pages.
static usize num_free_pages = 0;	// Amount of unused pages.
static usize last_page = 0;			// The last page marked as used.

void pm_init(void* phys_base, PhysMemory* mem_map, usize num_entries)
{
	lock = spin_new();
	phys_addr = phys_base;

	// Check for the highest usable physical memory address, so we know how much memory to allocate for the bitmap.
	usize highest = 0;
	for (usize i = 0; i < num_entries; i++)
	{
		// Only care about memory that we are able to own.
		if (mem_map[i].usage != PhysMemoryUsage_Free)
			continue;

		// Record the last byte of the current region if its address is the highest yet.
		const usize region_end = mem_map[i].address + mem_map[i].length;
		if (region_end > highest)
			highest = region_end;
	}

	num_pages = highest / CONFIG_page_size;
	const usize map_size = ALIGN_UP(num_pages / 8, CONFIG_page_size);

	// Get a memory region large enough to contain the bitmap.
	for (usize i = 0; i < num_entries; i++)
	{
		// Only care about memory that we are able to own.
		if (mem_map[i].usage != PhysMemoryUsage_Free)
			continue;

		if (mem_map[i].length >= map_size)
		{
			bit_map = phys_addr + mem_map[i].address;
			// The region where the bitmap is stored is inaccessible now.
			// * We could mark an entire page as used, but that would be wasteful.
			mem_map[i].address += map_size;
			mem_map[i].length -= map_size;
			break;
		}
	}

	// Mark all pages as used.
	memset(bit_map, 0xFF, map_size);

	for (usize i = 0; i < num_entries; i++)
	{
		// Only care about memory that we are able to own.
		if (mem_map[i].usage != PhysMemoryUsage_Free)
			continue;

		for (usize j = 0; j < mem_map[i].length; j += CONFIG_page_size)
		{
			// Mark the actual free pages as unused.
			bitmap_clear(bit_map, (mem_map[i].address + j) / CONFIG_page_size);
			num_free_pages++;
		}
	}

	arch_log("Initialized physical memory management, free memory = %u MiB\n",
			 (num_free_pages * CONFIG_page_size) / MiB);
}

void pm_update_phys_base(void* phys_base)
{
	// Get the physical address of the bit_map variable.
	PhysAddr bit_map_offset = (void*)bit_map - phys_addr;
	// Add the new offset back.
	bit_map = phys_base + bit_map_offset;
	// Update the physical base.
	phys_addr = phys_base;
}

void* pm_get_phys_base()
{
	return phys_addr;
}

static PhysAddr get_free_pages(usize amount, usize start)
{
	usize i = start;
	usize range_start = 0;

	// Get a region of consecutive pages that fulfill the requested amount.
	while (i < num_pages)
	{
		// If this page is used, skip it.
		if (bitmap_get(bit_map, i))
			goto next_page;

		range_start = i;

		// Otherwise, check if the next pages are free as well.
		// Start with the page after `i`.
		for (usize j = 1; j < amount; j++)
		{
			if (bitmap_get(bit_map, range_start + j))
				goto next_page;
		}

		// If we got here, that means we have found a region with `amount` consecutive pages.
		for (usize x = 0; x < amount; x++)
		{
			bitmap_set(bit_map, range_start + x);
		}

		last_page = range_start + amount + 1;
		return (PhysAddr)(range_start * CONFIG_page_size);

next_page:
		i++;
	}

	return 0;
}

PhysAddr pm_arch_alloc(usize amount)
{
	spin_acquire_force(&lock);

	PhysAddr mem = get_free_pages(amount, last_page);
	// If we couldn't find a free region starting at our last page offset, do another check, but from the beginning.
	// This is a lot slower, but a last resort because the other option is to panic as we are out of physical memory.
	if (mem == 0)
	{
		kassert(num_free_pages > 0, "Out of physical memory!");
		last_page = 0;
		mem = get_free_pages(amount, last_page);
	}

	kassert(mem != 0, "Unable to allocate %zu consecutive pages, total %zu available!", amount, num_free_pages);

	// Lastly, mark the pages as used now.
	num_free_pages -= amount;

	spin_free(&lock);
	return mem;
}

void pm_arch_free(PhysAddr addr, usize amount)
{
	spin_acquire_force(&lock);

	// Mark the page(s) as free.
	const usize page_idx = addr / CONFIG_page_size;
	for (usize i = page_idx; i < page_idx + amount; i++)
	{
		bitmap_clear(bit_map, i);
	}
	num_free_pages += 1;

	spin_free(&lock);
}

File: src/arch/x86_64/pm.rs
// Physical memory management

use alloc::{rc::Rc, sync::Arc};

use crate::{
    arch::{PhysAddr, PAGE_SIZE},
    boot::BootInfo,
    memory::pm::{CommonPhysManager, PhysMemoryUsage},
    misc::{align_up, bitmap::BitMap},
    thread::spin::SpinLock,
};
use core::{
    cell::{Cell, RefCell},
    ptr::null_mut,
};

pub struct PhysManager {
    /// Access lock for the manager.
    lock: SpinLock,
    /// Base address of memory region where physical pages are mapped 1:1
    phys_base: *mut u8,
    /// Bit map which stores whether or not a page is used.
    bit_map: BitMap,
    /// Amount of total installed pages.
    num_pages: usize,
    /// Amount of total free pages.
    num_free_pages: usize,
}

unsafe impl Sync for PhysManager {}
unsafe impl Send for PhysManager {}

static mut PMM: PhysManager = PhysManager {
    lock: SpinLock::new(),
    phys_base: null_mut(),
    bit_map: BitMap::new(),
    num_pages: 0,
    num_free_pages: 0,
};

impl CommonPhysManager for PhysManager {
    unsafe fn init(info: &mut BootInfo) {
        assert!(info.identity_base != 0, "HHDM base was NULL!");

        let pmm = unsafe { &raw mut PMM };

        unsafe {
            (*pmm).phys_base = info.identity_base as *mut u8;
        }

        // Check for the highest usable physical memory address, so we know how much memory to allocate for the bitmap.
        let mut highest = 0;
        for entry in info.memory_map.iter() {
            // Only care about memory that we are able to own.
            if entry.usage != PhysMemoryUsage::Free {
                continue;
            }

            // Record the last byte of the current region if its address is the highest yet.
            let region_end = entry.address + entry.length as u64;
            if region_end > highest {
                highest = region_end;
            }
        }

        // How big the bit map needs to be to store a bit for each pages.
        let map_size;
        unsafe {
            (*pmm).num_pages = highest as usize / PAGE_SIZE;
            map_size = align_up((*pmm).num_pages / 8, PAGE_SIZE);
        }

        // Get a memory region large enough to contain the bitmap.
        for entry in info.memory_map.iter_mut() {
            // Only care about memory that we are able to own.
            if entry.usage != PhysMemoryUsage::Free {
                continue;
            }

            if entry.length >= map_size {
                unsafe {
                    (*pmm).bit_map = BitMap::from_ptr(
                        PhysManager::phys_base().add(entry.address as usize),
                        map_size,
                    );
                }

                // The region where the bitmap is stored is inaccessible now.
                // We could mark an entire page as used, but that would be wasteful.
                entry.address += map_size as u64;
                entry.length -= map_size;
                break;
            }
        }

        // Mark all pages as used.
        unsafe {
            (*pmm).bit_map.fill(true);
        }

        // Mark the actual free pages as unused.
        // TODO: This is extremely slow! Fix me!
        for entry in info.memory_map.iter() {
            // Only care about memory that we are able to own.
            if entry.usage != PhysMemoryUsage::Free {
                continue;
            }

            for page in (entry.address as usize / PAGE_SIZE)
                ..((entry.address as usize + entry.length) / PAGE_SIZE)
            {
                unsafe {
                    (*pmm).bit_map.set(page, false);
                    (*pmm).num_free_pages += 1;
                }
            }
        }

        //println!("Initialized physical memory management, free memory = {} MiB\n",
        //		 (num_free_pages * PAGE_SIZE) / MIB);

        todo!();
    }

    unsafe fn alloc(num_pages: usize) -> PhysAddr {
        todo!();
    }

    unsafe fn free(addr: PhysAddr, num_pages: usize) {
        // TODO:
        // Check if all pages in the given space are used. If not, that means the values given are gibberish.
        // Clear the bits for the respective pages in the bit map.
        todo!();
    }

    unsafe fn phys_base() -> *mut u8 {
        unsafe {
            return PMM.phys_base;
        }
    }
}
